File: merge-sort/mergeSort.js
/*
 * Recursive
 */
function merge(a,b) {
  if (a.length === 0 && b.length > 0) return b;
  if (b.length === 0 && a.length > 0) return a;

  const [ah, ...at] = a;
  const [bh, ...bt] = b;

  if (ah < bh) return [ah, ...merge(at, b)]
  else return [bh, ...merge(a, bt)]
}

function rec_msort(array) {
  if (array.length < 2) return array;

  const length = array.length / 2;
  const left = array.slice(0,length);
  const right = array.slice(length);

  return merge(rec_msort(left), rec_msort(right))
}

File: merge-sort/mergeSort.kt
fun <T:Comparable<T>> merge(a: List<T>, b: List<T>): List<T> {
    if (a.isEmpty() && b.isNotEmpty()) return b
    if (b.isEmpty() && a.isNotEmpty()) return a

    val (ah, at) = Pair(a[0], a.drop(1))
    val (bh, bt) = Pair(b[0], b.drop(1))

    return when {
        ah >= bh -> listOf(bh) + merge(a, bt)
        else -> listOf(ah) + merge(at, b)
    }
}

fun <T:Comparable<T>> mergeSort(list: List<T>): List<T> {
    val len = list.count()
    if (len < 2) return list

    val half = kotlin.math.ceil((len.toDouble() / 2)).toInt()
    val (left, right) = list.chunked(half)

    return merge(mergeSort(left), mergeSort(right))
}

fun main() {
    val sorted = mergeSort(listOf(6, 5, 2, 3, 0, 1))
    println("sorted: $sorted")
}
